pkg/appconsts/versioned_consts.go

```diff
@@ -0,0 +1,45 @@
+package appconsts
+
+import (
+	"fmt"
+
+	v1 "github.com/celestiaorg/celestia-app/pkg/appconsts/v1"
+)
+
+const (
+	LatestVersion = v1.Version
+)
+
+// SubtreeRootThreshold works as a target value for the number of subtree roots in the
+// share commitment. If a blob contains more shares than this number, then the height
+// of the subtree roots will gradually increase so that the amount remains within that limit.
+// The rationale for this value is described in more detail in ADR013
+// (./docs/architecture/adr-013).
+// ADR013 https://github.com/celestiaorg/celestia-app/blob/e905143e8fe138ce6085ae9a5c1af950a2d87638/docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md //nolint: lll
+func SubtreeRootThreshold(version uint64) int {
+	switch version {
+	case v1.Version:
+		return v1.SubtreeRootThreshold
+	default:
+		panic(unsupportedVersion(version))
+	}
```

Comment:
[question]
since this can't handle v2, how are we wanting to handle these function when introduce new verions? are we going to update them with each new version here? use `> and <`?

Reply:
I would be okay here with just returning the latest version until we actually change the constant, although I can see the value of literally forcing us to think about every constant when we hardfork.

Reply:
Yeah, if a constant remains the same over versions, I guess we can do:

```go
switch version {
case 1, 2, 3:
    // do something
case 4, 5:
    // do something else
default:
    panic()
}
```

Reply:
I feel like this could end up biting us, since with each upgrade we have to ensure that we have tested each of these to handle the new version. That's not that difficult, we just have to be super careful never to forget one on accident.

is there a value in panicking on an unexpected version? if a node doesn't upgrade, then tendermint will not accept an upgraded header as valid, so we should only ever hit that panic if we make developer errors.

the second reason I think we should consider something else is because this is arguably more difficult to read at a glance. Instead of defining each version for each constant, we simply define the changes

wdyt about something similar to
```go
switch version {
case version <= 3:
   return v1.SubtreeRootThreshold
case version > 3:
    return v2.SubtreeRootThreshold
}
```

this example is a bit contrived and we could just use an if/else or whatever, but yeah

Reply:
I don't think it will be hard to forget; the node panics when any of these aren't updated which means all our integration tests would break until we had updated every constant.

I do get that it may be laborious if every time we have to bump the major version, we need to duplicate all the constants so I don't mind using ranges (although we shouldn't prefix the constants with `v1` and `v2` because that's no longer true). 

I'm not sure whether we should be composing the logic to commit values for future versions. It really depends on how we handle upgrades. If we have other logic that prevents us from ever processing a block from a future version then we're okay, if we just follow what version consensus agrees upon then we could reach a situation where we're processing blocks different from others because we didn't upgrade.

I think I still find panicking safer

Reply:
> the node panics when any of these aren't updated which means all our integration tests would break until we had updated every constant.

The issue as far as I understand it is that there is no guarantee that we are calling these functions at every height. So using these to halt nodes that don't upgrade feels like a giant footgun, because I think it means we have to have tests that ensure that we are actually calling each of these functions for each version.

Since we want nodes to halt at a height, then we should always halt if there is consensus over a header with an unexpected app version no matter what. Fortunately, that is what we're doing now, and I don't think we should ever change that due to the above. 

https://github.com/celestiaorg/celestia-core/blob/34cd5ec4435ae404dd5f576a73c91481883291fd/state/validation.go#L21-L28

Since that's the case (also the case for state/block sync), AND if it is actually the case that there are no guarantees that we call each of these functions each block, _then_ I think we should instead rely on that and not rely on the network halting because something slipped past our tests

Reply:
Yes if users forget to upgrade they shouldn't hit this panic anyway (they should be warned of an app version change prior). 

I guess our debate comes down to a difference in confidence in detection. I am confident, especially as it is called every height, that this would be detected in our testing and would prefer that we detect it easily (by panicking) then failing to detect it because it silently continues. You, if I understand correctly, have less confidence that we would detect such a scenario and would rather (if it somehow bypassed the initial check) that it continued to operate silently, however changing these variables across versions would likely cause an apphash mismatch causing a halt in any case. As an example, say we adjusted the `SubtreeRootThreshold` across versions, two different binaries would have different thresholds for the version and thus we'd see a data root mismatch. So for both cases, in the worst case that we bypass the app version check in core, would cause a halt. 

I guess exploring both paths, I am actually not too bothered either way (it's relatively insignificant in impact). I would still prefer the panic. I can also write some defensive tests i.e.

```go
for v := 1, v <= latestVersion; v++ {
    assert.Doesn'tPanic(t, variable(v))
}
```

Reply:
While we are calling each of them now at each height, but are there are scenarios in which that is not the case? For example, in the mint module when we change the inflation rate once a year?

I actually disagree that it has relatively insignificant impact :slightly_smiling_face:

If and only if we remove a consensus time check that afaiu we don't want to remove, we loose an marginally improved UX for people who have not upgraded. However, we gain not having to worry about human error that could halt the chain. While we aren't going to make that mistake now, but what about in 10 years?

I'm not going to block on it because all of the constant are getting called, but tbh this tradeoff does not seem worth it to me.

Reply:
> If and only if we remove a consensus time check

Why would we remove the validation check of the application version? We may also want to add an additional validation check in something like `BeginBlock` where we assert that the version never exceeds the latest version that the binary can handle - just as a sanity check.

> However, we gain not having to worry about human error that could halt the chain

As I explained above, not panicking would in most situations also result in halting the chain due to a data root mismatch or prepare/process proposal inconsistency (because nodes were using different values to calculate the subtreeroot threshold as an example).  At least with panicking we know instantly where we fucked up.

Happy to sync on this

Reply:
Just synced in person. The decision is to allow support for future versions i.e. the v1 SubtreeRootThreshold will be used for all versioning numbers until a change is introduced. 

This does not affect upgrades as other logic ensures that the state machine only processes blocks it can handle